src/lib/marshal/email-layout/writer.ts

```diff
@@ -0,0 +1,193 @@
+): AnyObj => {
+  // Move read only field under the dedicated field "__readonly".
+  const readonlyFields = emailLayout.__annotation?.readonly_fields || [];
+  const [readonly, remainder] = split(emailLayout, readonlyFields);
+  const emailLayoutjson = { ...remainder, __readonly: readonly };
+
+  // Strip out all schema annotations, so not to expose them to end users.
+  return omitDeep(emailLayoutjson, ["__annotation"]);
+};
+
+/* Builds an email layout dir bundle, which consist of the email layout JSON + the extractable files.
+   Then writes them into a layout directory on a local file system.
+*/
+export const writeEmailLayoutDirFromData = async (
+  emailLayoutDirCtx: LayoutDirContext,
+  emailLayout: EmailLayoutData<WithAnnotation>,
+): Promise<void> => {
+  const backupDirPath = path.resolve(sandboxDir, uniqueId("backup"));
+  const bundle = buildEmailLayoutDirBundle(emailLayout);
+
+  try {
+    // We store a backup in case there's an error.
+    if (emailLayoutDirCtx.exists) {
+      await fs.copy(emailLayoutDirCtx.abspath, backupDirPath);
+      await fs.emptyDir(emailLayoutDirCtx.abspath);
+    }
+
+    const promises = Object.entries(bundle).map(([relpath, fileContent]) => {
+      const filePath = path.resolve(emailLayoutDirCtx.abspath, relpath);
+
+      return relpath === LAYOUT_JSON
+        ? fs.outputJson(filePath, fileContent, { spaces: DOUBLE_SPACES })
+        : fs.outputFile(filePath, fileContent);
+    });
+
+    await Promise.all(promises);
+  } catch (error) {
+    // In case of any error, wipe the target directory that is likely in a bad
+    // state then restore the backup if one existed before.
+    if (emailLayoutDirCtx.exists) {
+      await fs.emptyDir(emailLayoutDirCtx.abspath);
+      await fs.copy(backupDirPath, emailLayoutDirCtx.abspath);
+    } else {
+      await fs.remove(emailLayoutDirCtx.abspath);
+    }
+
+    throw error;
+  } finally {
+    // Always clean up the backup directory in the temp sandbox.
+    await fs.remove(backupDirPath);
+  }
+};
+
+/* For a given email layout payload, this function builds a "email layout directoy bundle".
+   This is an object which contains all the relative paths and its file content.
+   It includes the extractable fields, which are extracted out and added to the bundle as separate files.
+*/
+const buildEmailLayoutDirBundle = (
+  emailLayout: EmailLayoutData<WithAnnotation>,
+): EmailLayoutDirBundle => {
+  const bundle: EmailLayoutDirBundle = {};
+
+  // A map of extraction settings of every field in the email layout
+  const compiledExtractionSettings = compileExtractionSettings(emailLayout);
+
+  // Iterate through each extractable field, determine whether we need to
+  // extract the field content, and if so, perform the
+  // extraction.
+  for (const [objPath, extractionSettings] of compiledExtractionSettings) {
+    const objPathStr = ObjPath.stringify(objPath);
+    const { default: extractByDefault, file_ext: fileExt } = extractionSettings;
+
```

Comment:
Related to the other comment, here we want to read the local email layout and look up the file path used in that layout.json file, if available, and use that file path. You can check out the example in the workflow writer module.